Add audio-to-text pipeline

eliteprox opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->
Adds the new `/audio-to-text` pipeline to go-livepeer, supporting the `openai/whisper-large-v3` model.

File formats supported are mp3, m4a, mp4, webm, and flac

This change requires https://github.com/livepeer/ai-worker/pull/103 and https://github.com/livepeer/lpms/pull/407

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- Refactors `handleAIRequest` and `processAIRequest` to support new response types like TextResponse
- Adds `/audio-to-text` endpoint to `ai_mediaserver.go`
- Pricing fixed to one pixel per millisecond
- Uses `ffmpeg.GetCodecInfo` to calculate duration and requires the lpms pull request above


**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

- Tested with rich vocal audio up to 4 hours long. Regression tested other pipelines to ensure refactoring cause any issues.
- Tested with all supported file formats and unsupported ones

curl request example:
```
curl --request POST   --url http://dev.eliteencoder.net:8937/audio-to-text --header 'Content-Type: multipart/form-data'   --form 'audio=@Recording.mp3'   --form 'model_id=openai/whisper-large-v3'   --form seed=123
```

**Does this pull request close any open issues?**
<!-- Fixes # -->
LIV-429
LIV-289

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./CONTRIBUTING.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

eliteprox committed:
Add speech-to-text pipeline, refactor processAIRequest and handleAIRequest to allow for various response types

eliteprox committed:
Pin gomod to ai-runner for testing

eliteprox committed:
Revert "Pin gomod to ai-runner for testing"

This reverts commit d4ba500dda02431f3627d7979d01e299c071bdec.

eliteprox committed:
Update go mod dep for ai-worker

eliteprox committed:
Calculate pixel value of audio file

eliteprox committed:
fix go-mod deps

eliteprox committed:
Adjust price calculation

eliteprox committed:
one second per pixel

eliteprox committed:
cleanup, fix missing duration

eliteprox committed:
Add supported file types, calculate price by milliseconds

eliteprox committed:
Add bad request response for unsupported file types

eliteprox committed:
Update name of function

eliteprox commented:
Added error handlers to respond with "400 bad request" when duration cannot be calculated due to unsupported file format or file corruption. This prevents invalid jobs from being sent to the network.

eliteprox committed:
Update go mod to ai-runner

emranemran commented on common/util.go:
```diff
@@ -530,3 +538,130 @@ func ParseEthAddr(strJsonKey string) (string, error) {
+var ErrUnsupportedFormat = errors.New("Unsupported audio file format. Supported formats: mp3, wav, mp4, m4a, webm, flac")
+var ErrorCalculatingDuration = errors.New("Error calculating duration")
+
+func CalculateAudioDuration(audio types.File) (int64, error) {
```
Would it make sense to just use ffmpeg to get the audio track's duration instead of using multiple methods based on fileExt? We do similar things for the video track in catalyst-api. You wouldn't need to do everything we do in catalyst-api [1] but just a call to the the go ffmpeg package might suffice. It would be less code to maintain. 

[1] https://github.com/livepeer/catalyst-api/blob/main/video/probe.go#L93

ad-astra-video replied:
This requires ffprobe installed right? I like this idea but it does add a dependency to using go-livepeer binaries.

The LPMS way would be to use this update to [GetCodecInfo](https://github.com/livepeer/lpms/pull/376) to get the duration.  With the ffmpeg upgrade to 7.0, should be covered on codec support as well.

eliteprox replied:
@emranemran We have updated the duration calculation to use ffmpeg. It required a change to [install_ffmpeg.sh](https://github.com/livepeer/go-livepeer/pull/3078/files#diff-4ae778054809274731b9da0c6a5a869c0bd214e92f954a5c9c39181748c2f175) and lpms https://github.com/livepeer/lpms/pull/407

rickstaa replied:
Great! thanks @eliteprox.

emranemran commented on server/ai_process.go:
```diff
@@ -421,14 +507,25 @@ func processAIRequest(ctx context.Context, params aiRequestParams, req interface
 	}
 
-	var resp *worker.ImageResponse
+	//var resp *worker.ImageResponse
```
minor: remove comment?

eliteprox replied:
Yes, cleaned up

emranemran commented:
Overall LGTM. I would recommend using ffmpeg for audio track processing (like calculating durations) and *only* accepting audio input instead of videos -- ffmpeg can help with that as well and we could potentially commonize the `Probe` package I linked in catalyst-api in the comments. Also, we need to understand what the file length limits are like.

eliteprox committed:
Use ffmpeg to get duration

ad-astra-video committed:
update install_ffmpeg.sh to parse audio better

eliteprox committed:
Check for audio codec instead of video codec

eliteprox committed:
gomod edits

eliteprox committed:
add docker file

leszko commented on common/util.go:
```diff
@@ -11,21 +12,28 @@ import (
+	"github.com/abema/go-mp4"
+	"github.com/ebml-go/webm"
 	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/go-audio/wav"
```
Nit: Have you considered using https://github.com/gopxl/beep? Seems to convert more formats + have more GitHub stars.

eliteprox replied:
We're going to drop support for WAV right now as we now have all other formats working through lpms/ffmpeg.

eliteprox replied:
WAV support re-added via ffmpeg

leszko commented on common/util.go:
```diff
@@ -530,3 +538,130 @@ func ParseEthAddr(strJsonKey string) (string, error) {
+	}
+}
+
+func GetDuration_WAV(audio io.ReadCloser) (int64, error) {
```
I think that all these functions would benefit from unit tests.

eliteprox replied:
Removed functions, writing tests is in the backlog

leszko commented on server/ai_process.go:
```diff
@@ -382,10 +400,77 @@ func submitUpscale(ctx context.Context, params aiRequestParams, sess *AISession,
+	}
+
+	// TODO: Refine this rough estimate in future iterations
+	sess.LatencyScore = took.Seconds() / float64(outPixels)
```
I'd consider extracting all these "temp" conversions hacks (latency score, outPixels) to one place. Otherwise it may be hard later to reason about them.

eliteprox replied:
Added a linear issue to track this https://linear.app/livepeer-ai/issue/AI-505/refactor-latency-scoring-into-common-function

leszko reviewed (approved):
Added some comments, other than that LGTM

eliteprox committed:
Update install_ffmpeg.sh to improve audio support, Add duration validation and logging, pin lpms

eliteprox committed:
rename speech-to-text to audio-to-text

eliteprox committed:
Update go-mod

eliteprox committed:
cleanup

eliteprox committed:
Merge branch 'add-speech-to-text' into rename-audio-to-text

eliteprox committed:
Merge pull request #3 from eliteprox/rename-audio-to-text

Rename audio to text

eliteprox committed:
update go mod

eliteprox committed:
remove comment

eliteprox committed:
Merge branch 'ai-video' into add-speech-to-text

eliteprox committed:
update gomod

eliteprox committed:
Update lpms mod

eliteprox committed:
Update to latest lpms

eliteprox committed:
Update lpms

rickstaa commented on go.sum:
```diff
@@ -182,6 +182,8 @@ github.com/dop251/goja_nodejs v0.0.0-20210225215109-d91c329300e7/go.mod h1:hn7BA
 github.com/dop251/goja_nodejs v0.0.0-20211022123610-8dd9abb0616d/go.mod h1:DngW8aVqWbuLRMHItjPUyqdj+HWPvnQe8V8y1nDpIbM=
 github.com/dustin/go-humanize v1.0.1 h1:GzkhY7T5VNhEkwH0PVJgjz+fX1rhBrR7pRT3mDkpeCY=
 github.com/dustin/go-humanize v1.0.1/go.mod h1:Mu1zIs6XwVuF/gI1OepvI0qD18qycQx+mFykh5fBlto=
+github.com/eliteprox/ai-worker v0.0.0-20240705062703-0908b518eb12 h1:LCnOJCD97i2jmR7gY5gJieYLS+01XHnbWkCpmHC1t28=
```
Quick note to replace when worker code is merged.

rickstaa commented on go.mod:
```diff
@@ -205,3 +206,5 @@ require (
 	rsc.io/tmplfunc v0.0.3 // indirect
 )
+
+replace github.com/livepeer/ai-worker => github.com/eliteprox/ai-worker v0.0.0-20240705062703-0908b518eb12
```
Remove when releasing.

rickstaa commented on go.sum:
```diff
@@ -531,16 +533,14 @@ github.com/libp2p/go-netroute v0.2.0 h1:0FpsbsvuSnAhXFnCY0VLFbJOzaK0VnP0r1QT/o4n
 github.com/livepeer/livepeer-data v0.7.5-0.20231004073737-06f1f383fb18/go.mod h1:Jpf4jHK+fbWioBHRDRM1WadNT1qmY27g2YicTdO0Rtc=
-github.com/livepeer/lpms v0.0.0-20240120150405-de94555cdc69 h1:4A6geMb+HfxBBfaS24t8R3ddpEDfWbpx7NTQZMt5Fp4=
-github.com/livepeer/lpms v0.0.0-20240120150405-de94555cdc69/go.mod h1:Hr/JhxxPDipOVd4ZrGYWrdJfpVF8/SEI0nNr2ctAlkM=
+github.com/livepeer/lpms v0.0.0-20240711175220-227325841434 h1:E7PKN6q/jMLapEV+eEwlwv87Xe5zacaVhvZ8T6AJR3c=
```
Ensure versioning is correct.

rickstaa commented on core/orchestrator.go:
```diff
@@ -945,6 +949,16 @@ func (n *LivepeerNode) upscale(ctx context.Context, req worker.UpscaleMultipartR
 
+func (n *LivepeerNode) AudioToText(ctx context.Context, req worker.AudioToTextMultipartRequestBody) (*worker.TextResponse, error) {
+
+	resp, err := n.AIWorker.AudioToText(ctx, req)
```
@eliteprox based on the code in https://github.com/livepeer/ai-worker/blob/0908b518eb1203617a3a68625d3326679a636439/worker/worker.go#L253-L305 we could simply forward the return values without performing the extra error check right 🤔?

eliteprox replied:
Removed unnecessary logic

rickstaa committed:
feat(ai): apply code improvements to AudioToText pipeline

This commit applies several code improvements to the AudioToText
codebase.

rickstaa commented on common/util.go:
```diff
@@ -530,3 +531,28 @@ func ParseEthAddr(strJsonKey string) (string, error) {
 }
+
+// determines the duration of an mp3 audio file by reading the frames
+var ErrUnsupportedFormat = errors.New("Unsupported audio file format")
```
@eliteprox. As also stated in https://github.com/eliteprox/go-livepeer/pull/4/files#r1677148579 this is a very useful error which should be implemented but currently is not used. I tried inputting the following unsupported audio OGG filetype:

[25eb6f27.zip](https://github.com/user-attachments/files/16227789/25eb6f27.zip)

But it threw a audio duration calculation error. Can we implement a audio file type unsupported error?

rickstaa commented:
@eliteprox looks like our pipeline fails to detect the duration of the following file:

[speech.zip](https://github.com/user-attachments/files/16227820/speech.zip)

Do you maybe know why 🤔?

eliteprox committed:
Remove unnecessary logic

eliteprox committed:
Merge branch 'add-speech-to-text' into add-speech-to-text-code-improvements

eliteprox committed:
Merge pull request #4 from eliteprox/add-speech-to-text-code-improvements

feat(ai): apply code improvements to AudioToText pipeline

eliteprox committed:
Remove unused error

eliteprox commented:
> @eliteprox looks like our pipeline fails to detect the duration of the following file:
> 
> [speech.zip](https://github.com/user-attachments/files/16227820/speech.zip)
> 
> Do you maybe know why 🤔?

This one appears to be a concatenated file and ffmpeg has issues calculating the duration in this case. The recommended solution is to re-encode the input to a consistent output format like flac. This can be combined with the effort to send audio-only to the ai-worker to optimize the pipeline.

rickstaa reviewed (approved):
LGTM 🚀.

eliteprox committed:
Fix missing err

eliteprox committed:
Update go.mod and tidy

rickstaa committed:
chore(ai): update ai-worker and lpms to latest version

This commit ensures that the ai-worker and lpms are at the latest
versions which contain the changes needed for the audio-to-text
pipeline.
